### ISSUE TITLE
HandlerList Memory Leak

### ISSUE THREAD

**Issue by SenseiKiwi** [YourGameSpace/TubeTils#1]
I observed a memory leak in AntiCooldown 4.0.3. A lot of fiddling with profilers and heap dumps revealed a buildup of `org.bukkit.event.HandlerList` instances which themselves hold references to other data.

More details:
- The server is running the latest build of Paper 1.16.5.
- The issue is not present in AntiCooldown 3.6.3.
- I reproduced the issue on a test server by having a player attack endermen. The longer they did this, the more instances of HandlerList accumulated, and the more memory they retained.
- The leak did not manifest on the same server with no players online over a ~68-hour period.
- It also didn't worsen when the test player stood around without attacking anything for a few minutes.

I have a theory for the underlying cause:
TubeTils defines events with _non-static_ HandlerList instances. All handler lists are registered to a global list owned by Bukkit. It seems they persist for the entire lifespan of the server. If you instantiate your events and they create new handler lists every time, then the handler lists will accumulate and any references they are retaining will be held in memory indefinitely.

The issue quickly gets out of hand on a server with 50+ players. A fix would be appreciated.

**Comment by TUBEOF** [comment]
Are there any kind of graphs in the sense of this that show the issue you can share?
Is it the attack hits themselves where the memory leak grows or is it exponential over time when players are on the server?

**Comment by SenseiKiwi** [comment]
I don't have graphs. I generated heap summaries with Spark at various times. I believe that the main increase is caused by players attacking, although there could be other things that cause smaller contributions.

Initially, I had a clue that the issue was related to AntiCooldown because we updated it just before the problems began. I stored a heap dump and the results showed a lot of memory being held by HandlerLists. The problem goes away when I downgrade to AntiCooldown 3.6.3, and it returns when I upgrade again. On the test server, HandlerLists barely used up any memory while we had the server running for 68 hours. I stayed AFK for a few hours with no effect. It was empty during the rest of the time. Then I asked someone to come on and stab mobs at a farm for ~50 minutes. We ended up with 1,188,339 HandlerList instances. Finally, I asked them to AFK for 5 minutes. The number of instances barely moved, to 1,188,345 (+6).

For reference, the declarations I'm referring to are like this one:
https://github.com/YourGameSpace/TubeTils/blob/647ddac6877db8a46f6e3d36124f8d624fa70377/src/main/java/de/tubeof/tubetils/api/cache/events/CacheContainerAddEvent.java#L8

This pattern is used in all of TubeTils' event classes. Each time one of these events is created, it will create a HandlerList instance which Bukkit will hold in its list of all handlers. It will then live on until the server shuts off. The issue would be avoided if `handlers` were static. One could argue that even if this is intended, it probably introduces other event-related bugs since Bukkit seems to expect one HandlerList per event class.
